add benchmarks/u_rho

zsweger opened the pull request:
### Briefly, what does this PR introduce?


### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No breaking changes.

### Does this PR change default behavior?
This PR should create an automatic benchmark using u-channel rho production to benchmark the tracking performance of the B0 trackers

Commit:
add benchmarks/u_rho

zsweger commented:
- I would like to make this into one of the automatic benchmarks which runs full simulations when there are detector or tracking updates, and then makes the plots which I've written here.
- What I've written now runs over reconstruction files from one of the simulation campaigns. 
- Can the experts please help me dit this benchmark to automatically run the simulations and then process the results?
- Also I have the benchmark produce some plots in the u_rho/figures directory. How can I convert these into artifacts?

veprbl commented:
Current `config.yml` is suitable for detector_benchmarks, but not physics benchmarks:
https://eicweb.phy.anl.gov/EIC/benchmarks/physics_benchmarks/-/pipelines/78788
<img width="655" alt="image" src="https://github.com/eic/physics_benchmarks/assets/245573/385b4150-bf76-4e13-8301-87862c20c00c">
You'd need to look at https://github.com/eic/physics_benchmarks/blob/master/benchmarks/diffractive_vm/config.yml for reference, but it's at least a matter of replacing `.det_benchmark` with `.phy_benchmark`.

veprbl commented:
Producing artifacts requires placing your output files to results directory. I think I haven't implemented this for diffractive_vm yet. Ideas are welcome.

zsweger committed:
Update config.yml for phy benchmark

zsweger commented:
The config.yml file has been updated for the physics benchmarks

zsweger commented:
> Producing artifacts requires placing your output files to results directory. I think I haven't implemented this for diffractive_vm yet. Ideas are welcome.

Sorry for the naïve questions, but does this just mean I should rename the `figures` directory to `results`?

zsweger committed:
Updated config.yml to include results task

zsweger committed:
Update config.yml

Updated config.yml to include simulate task

veprbl commented:
> > Producing artifacts requires placing your output files to results directory. I think I haven't implemented this for diffractive_vm yet. Ideas are welcome.
> 
> Sorry for the naïve questions, but does this just mean I should rename the `figures` directory to `results`?

Yes.

veprbl commented on benchmarks/u_rho/config.yml:
```diff
@@ -0,0 +1,40 @@
+  stage: collect
+  needs: ["u_channel_rho:simulate"]
+  script:
+    - cp benchmark_output/campaign_23.12.0_combined_60files_eicrecon.edm4eic.plots_figures/*.pdf figures/
```
One issue here is that the contents of the "current" directory is not preserved, only what's in the $LOCAL_DATA_PATH space. I think it would be simpler to move this line to the line after you call snakemake, and let gitlab output `results/` directory as an artifact.

Commit:
Editing figures directory

zsweger commented:
Okay I edited the procedure to immediately create the results directory and copy to it.

Commit:
Edited the compile rule in the Snakefile to actually compile the script

zsweger committed:
Update config.yml

Testing removing analysis compilation from u-channel rho pipeline

zsweger committed:
Update config.yml

Reimplemented compilation step

zsweger committed:
Update Snakefile

Updated compile step in snakefile

zsweger committed:
Update Snakefile

Edited compile rule

zsweger committed:
Update Snakefile

Updated file paths

zsweger committed:
Update Snakefile

Updated compile path

zsweger commented:
@veprbl , @wdconinc  I don't know why this keeps failing at the compile step of the pipeline. Could you offer any insight?

veprbl committed:
Snakefile: include benchmarks/u_rho/Snakefile

veprbl committed:
benchmarks/u_rho/Snakefile: change compile rule to use common definition

veprbl commented:
Should be a matter of including into the top-level Snakefile. I've pushed the changes, let's see if it works.

veprbl commented:
Produces some pretty plots now: https://eicweb.phy.anl.gov/EIC/benchmarks/physics_benchmarks/-/jobs/2335547/artifacts/file/results/benchmark_rho_dNdu.pdf

zsweger commented:
Woah! Thanks for fixing that! I will edit a few more things and create a couple more artifacts and then request for the pull request to be merged.

zsweger commented:
Also, can someone offer guidance on how a benchmark could be made to give a yes or no flag? Like if the number of reconstructed rho mesons in the B0 is less than 5% of the samples for this benchmark, the code would flag that whatever changes were made were bad.

wdconinc commented:
In the ROOT script, print a useful message and return with a non- zero value.

veprbl commented:
> Also, can someone offer guidance on how a benchmark could be made to give a yes or no flag? Like if the number of reconstructed rho mesons in the B0 is less than 5% of the samples for this benchmark, the code would flag that whatever changes were made were bad.

There is a protocol to report the numerical values, which I haven't tried yet: https://eicweb.phy.anl.gov/EIC/benchmarks/common_bench and it's used occasionally https://github.com/eic/physics_benchmarks/blob/3869b5480db7507af13b9e8bedc8129074fff2a2/benchmarks/dis/analysis/dis_electrons.cxx#L792. We probably want to implement it fully at some point (the presentation part of it doesn't exist yet), so I would add metrics and your proposal sounds great.

Commit:
Added B0 reco QA plots

zsweger committed:
Attempt at loading common_bench

zsweger committed:
Attempt at setting benchmark status flag

zsweger committed:
Update json file name

zsweger committed:
changed benchmark status function from void to int

zsweger committed:
fixed benchmark status definition location

zsweger committed:
Update RiceStyle.h to include common_bench

zsweger committed:
Moved benchmark status flag to plotting macro

zsweger committed:
Removed status flag from analysis code

zsweger committed:
Removed unnecessary dependency from pleaseIncludeMe.h

zsweger commented:
@veprbl I think I've added a status flag for the benchmark based on the rho reconstruction efficiency falling below 90%. From what I've seen, the efficiency is typically 95% so it falling below 90% should raise some red flags about the B0 detector. I added the method to `plot_rho_physics_benchmark.C` as `setbenchstatus()`. I believe it makes a json file that reports the status now. I don't know how to make it so that the CI tools know to check for this status flag. Any thoughts?

zsweger committed:
Added pdf explaining various benchmark plots

zsweger committed:
Copy to plot explainer to artifacts

zsweger committed:
removed plot explainer from artifacts

veprbl commented:
The list of allowed keys seems to be given here https://eicweb.phy.anl.gov/EIC/benchmarks/common_bench/-/blob/master/bin/collect_benchmarks.py?ref_type=heads#L119-121

In that logic, I don't see that `value` is compared to `target`, only reported. I suppose, the benchmark should do the comparison itself and report a correct value for `result`.

zsweger commented:
> The list of allowed keys seems to be given here https://eicweb.phy.anl.gov/EIC/benchmarks/common_bench/-/blob/master/bin/collect_benchmarks.py?ref_type=heads#L119-121
> 
> In that logic, I don't see that `value` is compared to `target`, only reported. I suppose, the benchmark should do the comparison itself and report a correct value for `result`.

I think I've done this in the plotting macro: [https://github.com/eic/physics_benchmarks/blob/42525a64159b3796373bbaa96ca495b59c1ae37b/benchmarks/u_rho/macros/plot_rho_physics_benchmark.C#L20](url)
so that it compares the rho reconstruction efficiency against a 90% threshold and reports the status flag.

zsweger committed:
Test reading produced json file

zsweger commented:
I tested this by adding the line:
`cat *.jason`
to the Snakefile. The pipeline log showed this so it seems to have worked:
`
{
    "tests": [
        {
            "description": "u-channel rho->pi+pi- reconstruction efficiency when both pions should be within B0 acceptance",
            "name": "rho_reconstruction_efficiency",
            "quantity": "efficiency",
            "result": "pass",
            "target": "0.9",
            "title": "rho Reconstruction Efficiency for rho -> pi+pi- in the B0",
            "value": 0.9605459586265728,
            "weight": 1.0
        }
    ]
}
`

zsweger committed:
Testing benchmark failure status

zsweger committed:
Update plot_rho_physics_benchmark.C

Second attempt at failure status

veprbl commented on benchmarks/u_rho/Snakefile:
```diff
@@ -0,0 +1,91 @@
+    input:
+        #lambda wildcards: [f"benchmark_output/campaign_23.12.0_{ix:04d}_eicrecon.edm4eic/plots.root" for ix in range(int(wildcards.NUM_FILES))],
+        lambda wildcards: expand(
+           "benchmark_output/campaign_23.12.0_{INDEX:04d}_eicrecon.edm4eic/plots.root",
```
Using campaign production is a good strategy for setting up. In the end, we will want to add the local production, so that when we test upstream software, we can see right away if it breaks this analysis.

zsweger replied:
Okay I've fixed this now so that users can set whether to use the simulation campaign or simulate things dynamically. The switch between these options is in `setup.config`

veprbl commented on flake.lock:
```diff
@@ -0,0 +1,224 @@
+{
```
Please remove this. I've accidentally committed those at the time, now they are removed.

zsweger replied:
Removed!

veprbl commented on flake.nix:
```diff
@@ -0,0 +1,46 @@
+{
```
This too.

zsweger replied:
Deleted!

veprbl commented on Snakefile:
```diff
@@ -18,3 +18,4 @@ root -l -b -q -e '.L {input}+'
 """
 
 include: "benchmarks/diffractive_vm/Snakefile"
+include: "benchmarks/u_rho/Snakefile"
```
Please, include the tex file for benchmarks/u_rho/BenchmarkPlotsExplained.pdf, maybe we can generate PDF on CI and store it artifacts along with the plots.

zsweger replied:
Okay I've now updated the latex file! I've tried the `pdflatex` command in eicshell to compile it and that didn't work. How can we generate the PDF on CI?

veprbl replied:
I've pushed a hacky solution. Perhaps, we need to request some small latex engine to be added to the container.

veprbl commented on benchmarks/u_rho/Snakefile:
```diff
@@ -0,0 +1,91 @@
+    shell:
+        """
+if [ ! -d "{input.plots}_figures" ]; then
+    mkdir "{input.plots}_figures"
```
Snakemake will pre-create all the directories for files you list in the `output`. I see there is a minor discrepancy between `edm4eic.plots.root_plots_figures` and `edm4eic.plots_figures`.

veprbl commented on benchmarks/u_rho/benchmark_output/README:
```diff
@@ -0,0 +1 @@
+This directory contains benchmark outputs
```
In principle we don't need this file. Snakemake will create all the directories.

zsweger replied:
Deleted!

veprbl commented:
This is very impressive progress! I've left a few comments above.

zsweger committed:
Copying json file to results directory

zsweger committed:
Update .gitlab-ci.yml to include u_rho

zsweger committed:
Update .gitlab-ci.yml

zsweger committed:
Update .gitlab-ci.yml

zsweger committed:
Update results collection in u_rho

zsweger committed:
Update snakefile json results

zsweger committed:
Test copying json from config.yml

zsweger committed:
Update json filename

zsweger committed:
Update .gitlab-ci.yml with u_rho results

zsweger committed:
remove results task in config.yml

zsweger committed:
Create benchmark.json

zsweger committed:
Update json file name

zsweger committed:
Update config.yml to have active results rule

zsweger committed:
Update config.yml test

zsweger committed:
Test commenting out diffractive_vm bench

zsweger committed:
Dont copy json to results

zsweger committed:
Move json to benchmark_outputs

zsweger committed:
Update snakefile for json path

zsweger committed:
Update results path

zsweger committed:
Move json file to results/u_rho

zsweger committed:
Update pipeline stages

zsweger committed:
Trigger pipeline again

zsweger committed:
Trigger pipelines

zsweger committed:
Create setup.config

zsweger committed:
Update pipelines with setup.config

zsweger committed:
Update setup.config path

zsweger committed:
While testing dont simulate all benchmarks

zsweger committed:
Update config.yml steps

zsweger committed:
Update config.yml test

zsweger committed:
Update Snakefile to get evgen files

zsweger committed:
Test grabbing evgen files from S3

zsweger committed:
Test not using simulation campaign

zsweger committed:
Fix Snakefile get evgen

zsweger committed:
Fix Snakefile again

zsweger committed:
Update Snakefile check

zsweger committed:
Fix snakefile check

zsweger committed:
Test detector simulation response

zsweger committed:
ddsim attempt 2

zsweger committed:
Create simulate.sh

zsweger committed:
Update config.yml with simulate.sh

zsweger committed:
Update setup.config

zsweger committed:
Fix simulate.sh

zsweger committed:
Update simulate.sh

zsweger committed:
Update simulate.sh --inputFiles

zsweger committed:
Update simulate.sh

zsweger committed:
Update config.yml to require common:detector

zsweger committed:
Test update config.yml

zsweger committed:
Test update config.yml again

zsweger committed:
Update config.yml

zsweger committed:
Update Snakefile input path

zsweger committed:
Update setup.config input path

zsweger committed:
Update config.yml tasks

zsweger committed:
Add simulate rule in Snakefile

zsweger committed:
Update config.yml with snakemake simulate rule

zsweger committed:
Fix Snakefile

zsweger committed:
Update Snakefile to refer to simulate.sh

zsweger committed:
Update simulate.sh

zsweger committed:
Update setup.config

zsweger committed:
Update config.yml reconstruction

zsweger committed:
Create reconstruct.sh

zsweger committed:
Update setup.config

zsweger committed:
Update config.yml

zsweger committed:
Update simulate.sh

zsweger committed:
Update config.yml

zsweger committed:
Update reconstruct.sh

zsweger committed:
Update setup.config

zsweger committed:
Update analyze rule

zsweger committed:
Update setup.config

zsweger committed:
Update setup.config

zsweger committed:
Create analyze.sh

zsweger committed:
Test getting mirrored file from XRootD

zsweger committed:
Update setup.config with sim_output

zsweger committed:
Update plot_rho_physics_benchmark.C with sim_output

zsweger committed:
Update Snakefile with sim_output

zsweger committed:
Update config.yml with sim_output

zsweger committed:
Trigger pipelines

zsweger committed:
Trigger pipelines again

zsweger committed:
Update analyze.sh with sim_output

zsweger committed:
Update config.yml with sim_output pdf path

zsweger committed:
Trigger pipelines

zsweger committed:
Update Snakefile paths test

zsweger committed:
Change config.yml with updated snakefile arg

zsweger committed:
Update reconstruct.sh rec file

zsweger committed:
Update `mkdir` to `mkdir -p` to avoid errors

zsweger committed:
Fix quotes in analyze.sh

zsweger committed:
Update quotes again in analyze.sh

zsweger committed:
Update figures path

zsweger committed:
Update config.yml figures path

zsweger committed:
Test using simulation campaign

zsweger committed:
Update Snakefile paths

zsweger committed:
Update config.yml paths

zsweger committed:
Update Snakefile paths

zsweger committed:
Update config.yml json path

zsweger committed:
Update Snakefile

zsweger committed:
Update plot_rho_physics_benchmark.C

zsweger committed:
Update Snakefile

zsweger committed:
Update config.yml paths

zsweger committed:
Update config.yml

zsweger committed:
Update config.yml path

zsweger committed:
Update config.yml test

zsweger committed:
Delete flake.lock

zsweger committed:
Delete flake.nix

zsweger committed:
Delete benchmarks/u_rho/benchmark_output/README

zsweger committed:
Test dynamic simulations again

zsweger committed:
Update analyze.sh json path

zsweger commented:
@veprbl I've added the ability to switch between running the benchmark over the simulation campaign files, and dynamically re-running the simulations. The switch is in `setup.config`. The pipeline seems to run, but the only issue now is that there seem to be no tracks that are actually reconstructed. Do you know why this might be? 

Also sorry for all the pushes, it's been a struggle testing getting the pipelines working!

zsweger committed:
Add latex files to explain the benchmark plots

zsweger committed:
Add format file for latex file

veprbl commented:
> @veprbl I've added the ability to switch between running the benchmark over the simulation campaign files, and dynamically re-running the simulations. The switch is in `setup.config`. The pipeline seems to run, but the only issue now is that there seem to be no tracks that are actually reconstructed. Do you know why this might be?

Everything looks right to me. Perhaps, it's diffractive events not often leaving tracks in the central detector?

veprbl committed:
compile manual with tectonic

veprbl committed:
use time from any path

zsweger committed:
Change location of latex compilation

I want the pdf file explaining each of the benchmark figures to be located in the same artifacts folder as the figures

zsweger commented:
> > @veprbl I've added the ability to switch between running the benchmark over the simulation campaign files, and dynamically re-running the simulations. The switch is in `setup.config`. The pipeline seems to run, but the only issue now is that there seem to be no tracks that are actually reconstructed. Do you know why this might be?
> 
> Everything looks right to me. Perhaps, it's diffractive events not often leaving tracks in the central detector?

You're right that these won't go into the central detector usually. The pions most often end up in the B0 tracker. But we should still be able to see them in the `ReconstructedChargedParticles` branch. From the simulation campaign we get plots like this:
<img width="861" alt="Screenshot 2024-05-10 at 1 48 41 PM" src="https://github.com/eic/physics_benchmarks/assets/29314865/45eada63-29bf-47ad-9094-912841c30daf">
But now I'm seeing this:
<img width="587" alt="Screenshot 2024-06-03 at 2 42 38 PM" src="https://github.com/eic/physics_benchmarks/assets/29314865/688e9639-5e6a-4553-aec0-a6de51ee6a2b">

zsweger committed:
Test 10k events

veprbl commented:
![image](https://github.com/eic/physics_benchmarks/assets/245573/9e1650b1-e4ae-4107-8fa3-670e613b7710)
Looks like it shows up with additional stats. We could produce an eSTARlight simulation sample with re-weighted events to more efficiently use detector simulation statistics.

zsweger commented:
> ![image](https://private-user-images.githubusercontent.com/245573/337386259-9e1650b1-e4ae-4107-8fa3-670e613b7710.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOi
[benchmark_rho_efficiencies.pdf](https://github.com/user-attachments/files/15686448/benchmark_rho_efficiencies.pdf)
JnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3MDgyNTAsIm5iZiI6MTcxNzcwNzk1MCwicGF0aCI6Ii8yNDU1NzMvMzM3Mzg2MjU5LTllMTY1MGIxLWU0YWUtNDEwNy04ZmEzLTY3MGU2MTNiNzcxMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNlQyMTA1NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMGQ5OWU1MzczZTc2Y2UwOTI0NmNmMDFkYjg2MTk0NDA2ZTY0NGY3NDVhNTQzNTFjOGFhZTA0NDA2YmFiZDcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.19_SUHmNaTYOVgGMosQmcUg_9VygbG4bjoTYEJe_oGM) Looks like it shows up with additional stats. We could produce an eSTARlight simulation sample with re-weighted events to more efficiently use detector simulation statistics.

In combination with this plot:
[benchmark_rho_efficiencies.pdf](https://github.com/user-attachments/files/15686451/benchmark_rho_efficiencies.pdf)
this makes me think that something changed about how reconstructed tracks from the B0 are stored, because it looks like only low-eta pions are being reconstructed, probably within the central trackers. I'll ask around!

zsweger commented:
Alex Jentsch just said that he's seen the same issue, that lately the B0 reconstruction has a problem

veprbl commented:
Understood. I've checked few things, but nothing stands out. Could you file this bug to https://github.com/eic/EICrecon/issues, please.

zsweger committed:
Update config.yml with May sim campaign

zsweger committed:
Update Snakefile with May sim campaign

zsweger committed:
Test use May sim campaign

zsweger committed:
Update config.yml with April campaign

zsweger committed:
Test April campaign

zsweger commented:
> Understood. I've checked few things, but nothing stands out. Could you file this bug to https://github.com/eic/EICrecon/issues, please.

Done! [https://github.com/eic/EICrecon/issues/1500](https://github.com/eic/EICrecon/issues/1500)

zsweger committed:
Test ReconstructedChargedRealPIDParticles

zsweger committed:
Test on live data

zsweger committed:
Test truth PID using chargedParticleAssociations

zsweger committed:
Test on 100 samples

zsweger committed:
Saeahram's fix to uchannelrho.cxx

zsweger committed:
Try 1000 events

zsweger committed:
Test simulation campaign

zsweger committed:
Update to July sim campaign

zsweger committed:
Update Snakefile to July campaign

zsweger committed:
Update z axis label sizes

zsweger committed:
July campaign has fewer files. Reduce to 45

zsweger committed:
Editing benchmark plot labels

zsweger committed:
Update EPIC label

zsweger committed:
Get rid of dN/dt plot

zsweger committed:
Test proton reco fix

zsweger committed:
Get rid of dN/dt plot second attempt

zsweger committed:
Update bench status test

zsweger committed:
Update bench plot labels

zsweger committed:
Update plots top margin

zsweger committed:
Final top margin fix

zsweger committed:
Test bench with 5k events

zsweger committed:
Hello California! (and John!)

Demonstration of benchmark for CA Consortium Meeting, using simulation campaign

zsweger committed:
Hello tutorial attendees!

zsweger committed:
Hello tutorial attendees! :)

zsweger committed:
Hello tutorial attendees!

zsweger committed:
Try to trigger pipelines

zsweger committed:
Trigger pipelines again

ShujieL committed:
re-trigger u-rho simulation
